### ISSUE TITLE
Keybase client automated restart after update vs systemd

### ISSUE THREAD

**Issue by rkujawa** [keybase/client#5355]
I wrote simple systemd units for keybase and kbfs, intended to be used as unprivileged user. This generally works okay and I can successfully manage these processes with systemd. 

However, if I have automatic updates enabled (via Keybase official RPM repository and using Fedora's dnf-automatic), keybase client gets restarted automatically as soon as I use the `keybase` command. The problem is that the original keybase processes managed by systemd terminates and the new keybase process gets started within user session. So from systemd point of view, the keybase process is dead now:

```
$ systemctl --user status keybase.service  
● keybase.service - Keybase
   Loaded: loaded (/home/rkujawa/.config/systemd/user/keybase.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Mon 2017-01-02 18:55:25 CET; 1 weeks 1 days ago
 Main PID: 747 (code=exited, status=0/SUCCESS)

Dec 15 07:36:10 vshell.home.c0ff33.net systemd[638]: Started Keybase.
```

But of course the service process is running, as it was restarted by the keybase command:
```
$ ps -fC keybase
UID        PID  PPID  C STIME TTY          TIME CMD
rkujawa  13290   638  0 Jan10 ?        00:01:14 /usr/bin/keybase --debug --log-file /home/rkujawa/.cache/keybase/...
```

What would be the best way to address this problem? Maybe the keybase service should restart itself, instead of relying on client command to do so? 

Currently, my systemd units are defined in the following way:

~/.config/systemd/user/keybase.service :
```
[Unit]
Description=Keybase

[Service]
ExecStart=/usr/bin/keybase -d --log-file=.cache/keybase/keybase.service.log service
PIDFile=/run/user/$UID/keybase/keybased.pid

[Install]
WantedBy=myscreen.target
```

~/.config/systemd/user/kbfs.service :
```
[Unit]
Description=Keybase kbfs
After=keybase.service

[Service]
ExecStart=/usr/bin/kbfsfuse -debug keybase 
ExecStopPost=/usr/bin/fusermount -uz keybase

[Install]
WantedBy=myscreen.target
```

**Comment by oconnor663** [comment]
I'm not familiar enough with systemd to have strong opinions about how things should be organized, but I think we'll want to add some first class support for it, as we've done with launchd on Mac. (Do a `grep -i launchd` to see some of what we've done there.) There are a few places where we'll want the client to speak systemd:

- restarting after updates, like you mentioned
- auto-forking, when a client command is run without the service having started yet (or maybe socket activation takes care of this?)
- logging to the journal instead of random files? (and the corresponding `keybase log send` command)
- autostart on boot (should the GUI app keep using `~/.config/autostart`?)

**Comment by IvanMalison** [comment]
@rkujawa Don't you just need to set the Type to forking?

**Comment by rkujawa** [comment]
@IvanMalison it's not an issue with processes started by systemd. It's an issue with the client, that upon keybase update, kills the service processes started by systemd, then starts them on its own. This way systemd loses track of these processes and thinks that service has failed. Setting service to forking does not solve that problem.

Ideally, like @oconnor663 wrote, client could talk to systemd in all cases where start/stop/restart of the service is needed. Or just use socket activation (someone already suggested in #3508). However supporting this functionality requires modification of the client and I didn't have the time to further investigate it yet.

Upon re-inspection... it seems that keybase service behaviour has changed recently. With systemd service configured still as in original post, now restart after upgrade just fails:
```
$ keybase ping
▶ WARNING Restarting after upgrade; service is running v1.0.32-20171002183401+28678b058, while v1.0.34-20171012160058+c282fb227 is available
▶ ERROR dial unix /run/user/1000/keybase/keybased.sock: connect: no such file or directory
$ systemctl --user status keybase.service
● keybase.service - Keybase
   Loaded: loaded (/home/rkujawa/.config/systemd/user/keybase.service; enabled; vendor preset: enabled)
   Active: inactive (dead) since Thu 2017-10-19 09:14:18 CEST; 46s ago
  Process: 757 ExecStart=/usr/bin/keybase -d --log-file=.cache/keybase/keybase.service.log service (code=exited, status=0/SUCCESS)
 Main PID: 757 (code=exited, status=0/SUCCESS)

Oct 05 08:00:07 vshell.home.c0ff33.net systemd[647]: Started Keybase.
```

Of course restarting it still brings it up correctly and it works later on.

```
$ systemctl --user start keybase.service
$ systemctl --user status keybase.service
● keybase.service - Keybase
   Loaded: loaded (/home/rkujawa/.config/systemd/user/keybase.service; enabled; vendor preset: enabled)
   Active: active (running) since Thu 2017-10-19 09:18:59 CEST; 2s ago
 Main PID: 1392 (keybase)
   CGroup: /user.slice/user-1000.slice/user@1000.service/keybase.service
           └─1392 /usr/bin/keybase -d --log-file=.cache/keybase/keybase.service.log service

Oct 19 09:18:59 vshell.home.c0ff33.net systemd[647]: Started Keybase.
$  keybase ping
▶ INFO API Server at https://api.keybase.io is up
```

**Comment by oconnor663** [comment]
This is actually back on the front burner for the next week or so -- I'll update this thread when we make changes.

**Comment by oconnor663** [comment]
Ok, we just landed systemd support in master, and it'll go out with the next Linux build. (Might be a few days, due to some unrelated GUI changes.) It's currently disabled by default, but if you set `KEYBASE_SYSTEMD=1` in your environment it'll turn on. After we feel confident that it's working in the wild, we'll switch that default to on. Here's a summary of what we've got:

- There are now three Keybase user services: `keybase.service`, `kbfs.service`, and `keybase.gui.service`.
- There are also two socket units, `keybase.socket` and `kbfs.socket`, binding to the same filesystem paths as now, referenced with `Requires=` dependencies in their respective service files. I considered not including these, since there are some funky cases where the sockets get deleted while systemd thinks they're still there, but for now they're in. Socket activation won't happen under normal circumstances, but it works in theory.
- The `run_keybase` script will now start those three services. It also has some compatibility logic to try to handle the case where you upgrade / turn on systemd support while services are already running. (I believe there are still some races here, if you switch systemd support on and off repeatedly, but hopefully this is a corner case.)
- Starting on boot hasn't changed. `~/.config/autostart/keybase_autostart.desktop` still calls `run_keybase`.
- Autoforking (starting the background service if it's not running, when you run any CLI command) is done with `systemctl --user start keybase.service`. Detecting the running service hasn't changed, however, so there shouldn't be any funky upgrade scenarios to worry about here.

If you have any existing unit files of the same name, you might want to rename them, to avoid shadowing these packaged ones when they ship.
